packages/babel-parser/src/plugins/typescript/index.js

```diff
@@ -2043,6 +2043,10 @@ export default (superClass: Class<Parser>): Class<Parser> =>
       state: N.ParseSubscriptState,
     ): N.Expression {
       if (!this.hasPrecedingLineBreak() && this.match(tt.bang)) {
+        // When ! is consumed as a postfix operator (non-null assertion),
+        // disallow JSX tag forming after. e.g. When parsing `p! < n.p!`
+        // `<n.p` can not be a start of JSX tag
+        this.state.exprAllowed = false;
```

Comment:
It was incorrectly removed in https://github.com/babel/babel/pull/13431/files#diff-e2c7279fd768044b26448340eef2ba1e8c62a5f36bf280621af540837fa5881dL2049